### PR TITLE
add optional Foreign encoding and decoding

### PR DESCRIPTION
**Pull request by shmish111**
`Foreign.Generic` seems to be the correct way to deal with JSON these days and in the process of upgrading a project from PS 0.11 to PS 0.12 I discovered that I needed these instances to be generated.

### CODE REVIEW

File: src/Language/PureScript/Bridge/Printer.hs [hunk 1]
```diff
@@ -61,13 +60,13 @@ moduleToText settings m = T.unlines $
   "-- File auto generated by purescript-bridge! --"
   : "module " <> psModuleName m <> " where\n"
   : map importLineToText allImports
-  ++ [ ""
-     , "import Prelude"
+  <> [ ""
+     , "import Prelude" -- TODO: need a switch for Foreign.Generic to see if we should import or not and then implementation of the instances
```
Comment (eskimor) [inline src/Language/PureScript/Bridge/Printer.hs]: Is that todo still relevant?

Comment (shmish111) [reply src/Language/PureScript/Bridge/Printer.hs]: this is why we review code 😄 

removed and pushed

**Comment by eskimor** [comment]
Thanks a lot! Looks good at first sight. Will need to do a more thorough review later on.

**Comment by eskimor** [comment]
Could you have a look at the failing test? And ideally add another one for the new functionality, this would also help the review.

**Comment by eskimor** [comment]
@shmish111 ?

**Comment by ford-prefect** [comment]
@shmish111 fwiw, I think one place this breaks is when dealing with sum types that reference other sum types.

You generate `GenericEncode` instances with `unwrapSingleConstructors = true` which is broadly okay, but in the nested case, `purescript-foreign-generic` explicitly callse `genericEncode` with that option set to `false`, and this breaks things (ref: https://github.com/paf31/purescript-foreign-generic/blob/v10.0.0/src/Foreign/Generic/Class.purs#L351).

Not sure how we can deal with this.

**Comment by shmish111** [comment]
Sorry @eskimor I lost track of this PR (even though we do use the fork).

@ford-prefect you can pass in `ForeignOptions` to change this, or am I missing something?

**Comment by ford-prefect** [comment]
> @ford-prefect you can pass in `ForeignOptions` to change this, or am I missing something?

@shmish111 that's a bit of a hammer (I guess we could do that and manually tweak the Haskell side instances to also not unwrap single constructors), but when set to true, it's basically not possible to use nested sum types.

**Comment by shmish111** [comment]
@ford-prefect (awesome name BTW) I'm not sure what would be the best default setup. All I can say is that I'm working on a real-world project and this default setup worked so I feel you would need to give a counter-example to do something different.

@eskimor fixed the test. I feel this project could do with a test that uses purescript to compile a file however I've been down the path of testing one language from another and it's not easy, not sure if it would be worth it or not. Perhaps it's possible using purescript as a haskell library to parse and type check? Anyway, not for this PR.

**Comment by eskimor** [comment]
Thanks a lot!

**Comment by ford-prefect** [comment]
@shmish111 I think it makes sense to keep what you have for now. If we find anything that addresses our problems in a general way, I'll make another PR for it. Thanks again for this work (to you and @eskimor)!

**Pull request merged**